apoc/_table_row_classifier.py

```diff
@@ -0,0 +1,207 @@
+        self.classifier.to_opencl_file(self.classifier.opencl_file, overwrite_classname=self.classifier_classname)
+
+    def predict(
+            self,
+            feature_table: Dict[str, Union[List[float], np.ndarray]],
+            return_numpy: bool = True
+    ) -> np.array:
+        """Predict row class from a table.
+
+        Parameters
+        ----------
+        feature_table : Dict[str, Union[List[float], np.ndarray]]
+            The table from which to make the prediction. Each row of the table
+            will be classified. The table can either be a pandas DataFrame or a
+            Dict with string keys (column names) and numpy array columns.
+        return_numpy : bool
+            If True, the resulting predictions are returned as a numpy array.
+            If False, the predictions are returned as a pyopencl array.
+            The default value is True.
+
+        Returns
+        -------
+        output : np.ndarray
+            An array containing the predicted class for each row.
+        """
+        import pyclesperanto_prototype as cle
+
+        ordered_features = self.order_feature_table(feature_table)
+
+        # allocate the result
+        output = cle.create_like(ordered_features[0].shape)
+
+        # push the features
+        parameters = {}
+        for i, f in enumerate(ordered_features):
+            parameters['in' + str(i)] = cle.push(f)
+        parameters['out'] = output
+
+        # run the classifier
+        cle.execute(None, self.classifier.opencl_file, "predict", ordered_features[0].shape, parameters)
+
+        if return_numpy is True:
+            return np.asarray(output, dtype=np.uint32)
+        else:
+            return output
+
+    def _prepare_feature_table(
+            self,
+            feature_table: Dict[str, Union[List[float], np.ndarray]]
+    ) -> List[np.ndarray]:
+        """Prepare a feature table for training.
+
+        This coerces the feature table into the form expected by the classifier
+        (list of numpy array) and stores the order of the features.
+
+        Parameters
+        ----------
+        feature_table : Dict[str, Union[List[float], np.ndarray]]
+            The table from which to make the prediction. Each row of the table
+            will be classified. The table can either be a pandas DataFrame or a
+            Dict with string keys (column names) and numpy array columns.
+
+        Returns
+        -------
+        ordered_features : List[np.ndarray]
+            The features stored in a list. The order of the features is
+            specified by self.ordered_feature_names
+        """
+        if len(self._ordered_feature_names) == 0:
+            # if the feature names haven't previously been stored,
+            # store them as a list
+            self._ordered_feature_names = list(feature_table.keys())
+            self.classifier.feature_specification = " ".join(self.ordered_feature_names)
```

Comment:
My understanding is that the `feature_specification` is a string of the column names, each one separated by a space. Is that correct?

Reply:
Yes! For backwards-compatibility, also comma-separated lists should work.